File: pkg/ui/src/api/services/sqs.ts
/**
 * SQS Service API Client
 * REST HTTP client for SQS via Go proxy
 * @module api/services/sqs
 */

import { PROXY_BACKEND } from '@/config'
import { APIError } from '../client'

const api = PROXY_BACKEND.replace(/\/$/, '')

/**
 * Extract queue name from a queue URL.
 * Queue URL format: https://sqs.region.amazonaws.com/account-id/QueueName
 */
function extractQueueName(queueUrl: string): string {
  return queueUrl.split('/').filter(Boolean).pop() || queueUrl
}

export class SQSService {
  async listQueues(prefix?: string): Promise<string[]> {
    const url = prefix
      ? `${api}/sqs/queues?prefix=${encodeURIComponent(prefix)}`
      : `${api}/sqs/queues`
    const res = await fetch(url)
    if (!res.ok) {
      const errorText = await res.text()
      throw new APIError(`List queues failed: ${errorText}`, res.status, 'sqs')
    }
    const data = await res.json()
    return data.QueueUrls || []
  }

  async getQueueUrl(queueName: string): Promise<string | undefined> {
    const res = await fetch(`${api}/sqs/queues/${encodeURIComponent(queueName)}`)
    if (!res.ok) {
      const errorText = await res.text()
      throw new APIError(`Get queue URL failed: ${errorText}`, res.status, 'sqs')
    }
    const data = await res.json()
    return data.QueueUrl
  }

  async getQueueAttributes(queueUrl: string, attributeNames: string[] = ['All']): Promise<Record<string, string>> {
    const queueName = extractQueueName(queueUrl)
    const query = new URLSearchParams()
    attributeNames.forEach(n => query.append('AttributeName', n))
    const res = await fetch(`${api}/sqs/queues/${encodeURIComponent(queueName)}/attributes?${query.toString()}`)
    if (!res.ok) {
      const errorText = await res.text()
      throw new APIError(`Get queue attributes failed: ${errorText}`, res.status, 'sqs')
    }
    const data = await res.json()
    const attributes: Record<string, string> = {}
    if (data.Attributes) {
      Object.entries(data.Attributes).forEach(([key, value]) => {
        attributes[key] = String(value)
      })
    }
    return attributes
  }

  async createQueue(queueName: string, options?: {
    Attributes?: Record<string, string>
    tags?: Record<string, string>
  }): Promise<{ QueueUrl: string }> {
    const res = await fetch(`${api}/sqs/queues`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ QueueName: queueName, ...options }),
    })
    if (!res.ok) {
      const errorText = await res.text()
      throw new APIError(`Create queue failed: ${errorText}`, res.status, 'sqs')
    }
    const data = await res.json()
    return { QueueUrl: data.QueueUrl || '' }
  }

  async deleteQueue(queueUrl: string): Promise<void> {
    const queueName = extractQueueName(queueUrl)
    const res = await fetch(`${api}/sqs/queues/${encodeURIComponent(queueName)}`, { method: 'DELETE' })
    if (!res.ok) {
      const errorText = await res.text()
      throw new APIError(`Delete queue failed: ${errorText}`, res.status, 'sqs')
    }
  }

  async sendMessage(queueUrl: string, messageBody: string, options?: {
    DelaySeconds?: number
    MessageAttributes?: Record<string, any>
    MessageDeduplicationId?: string
    MessageGroupId?: string
  }): Promise<{ MessageId: string; MD5OfMessageBody: string }> {
    const queueName = extractQueueName(queueUrl)
    const res = await fetch(`${api}/sqs/queues/${encodeURIComponent(queueName)}/messages`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ MessageBody: messageBody, ...options }),
    })
    if (!res.ok) {
      const errorText = await res.text()
      throw new APIError(`Send message failed: ${errorText}`, res.status, 'sqs')
    }
    const data = await res.json()
    return {
      MessageId: data.MessageId || '',
      MD5OfMessageBody: data.MD5OfMessageBody || '',
    }
  }

  async receiveMessage(queueUrl: string, options?: {
    MaxNumberOfMessages?: number
    WaitTimeSeconds?: number
    VisibilityTimeout?: number
    AttributeNames?: string[]
    MessageAttributeNames?: string[]
  }): Promise<any[]> {
    const queueName = extractQueueName(queueUrl)
    const params = new URLSearchParams()
    if (options?.MaxNumberOfMessages !== undefined) params.set('MaxNumberOfMessages', String(options.MaxNumberOfMessages))
    if (options?.WaitTimeSeconds !== undefined) params.set('WaitTimeSeconds', String(options.WaitTimeSeconds))
    if (options?.VisibilityTimeout !== undefined) params.set('VisibilityTimeout', String(options.VisibilityTimeout))
    if (options?.AttributeNames) params.set('AttributeNames', options.AttributeNames.join(','))
    if (options?.MessageAttributeNames) params.set('MessageAttributeNames', options.MessageAttributeNames.join(','))
    const qs = params.toString()
    const url = qs
      ? `${api}/sqs/queues/${encodeURIComponent(queueName)}/messages?${qs}`
      : `${api}/sqs/queues/${encodeURIComponent(queueName)}/messages`
    const res = await fetch(url)
    if (!res.ok) {
      const errorText = await res.text()
      throw new APIError(`Receive message failed: ${errorText}`, res.status, 'sqs')
    }
    const data = await res.json()
    return data.Messages || []
  }

  async deleteMessage(queueUrl: string, receiptHandle: string): Promise<void> {
    const queueName = extractQueueName(queueUrl)
    const res = await fetch(
      `${api}/sqs/queues/${encodeURIComponent(queueName)}/messages/${encodeURIComponent(receiptHandle)}`,
      { method: 'DELETE' }
    )
    if (!res.ok) {
      const errorText = await res.text()
      throw new APIError(`Delete message failed: ${errorText}`, res.status, 'sqs')
    }
  }

  async purgeQueue(queueUrl: string): Promise<void> {
    const queueName = extractQueueName(queueUrl)
    const res = await fetch(`${api}/sqs/queues/${encodeURIComponent(queueName)}/purge`, { method: 'POST' })
    if (!res.ok) {
      const errorText = await res.text()
      throw new APIError(`Purge queue failed: ${errorText}`, res.status, 'sqs')
    }
  }

  async setQueueAttributes(queueUrl: string, attributes: Record<string, string>): Promise<void> {
    const queueName = extractQueueName(queueUrl)
    const res = await fetch(`${api}/sqs/queues/${encodeURIComponent(queueName)}/attributes`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ Attributes: attributes }),
    })
    if (!res.ok) {
      const errorText = await res.text()
      throw new APIError(`Set queue attributes failed: ${errorText}`, res.status, 'sqs')
    }
  }

  async changeMessageVisibility(queueUrl: string, receiptHandle: string, visibilityTimeout: number): Promise<void> {
    const queueName = extractQueueName(queueUrl)
    const res = await fetch(`${api}/sqs/queues/${encodeURIComponent(queueName)}/visibility`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ ReceiptHandle: receiptHandle, VisibilityTimeout: visibilityTimeout }),
    })
    if (!res.ok) {
      const errorText = await res.text()
      throw new APIError(`Change message visibility failed: ${errorText}`, res.status, 'sqs')
    }
  }
}

export const sqsService = new SQSService()

export const listQueues = (prefix?: string) => sqsService.listQueues(prefix)
export const getQueueUrl = (queueName: string) => sqsService.getQueueUrl(queueName)
export const getQueueAttributes = (queueUrl: string, attributeNames?: string[]) =>
  sqsService.getQueueAttributes(queueUrl, attributeNames)
export const createQueue = (queueName: string, options?: Parameters<SQSService['createQueue']>[1]) =>
  sqsService.createQueue(queueName, options)
export const deleteQueue = (queueUrl: string) => sqsService.deleteQueue(queueUrl)
export const sendMessage = (queueUrl: string, messageBody: string, options?: Parameters<SQSService['sendMessage']>[2]) =>
  sqsService.sendMessage(queueUrl, messageBody, options)
export const receiveMessage = (queueUrl: string, options?: Parameters<SQSService['receiveMessage']>[1]) =>
  sqsService.receiveMessage(queueUrl, options)
export const deleteMessage = (queueUrl: string, receiptHandle: string) =>
  sqsService.deleteMessage(queueUrl, receiptHandle)
export const purgeQueue = (queueUrl: string) => sqsService.purgeQueue(queueUrl)
export const setQueueAttributes = (queueUrl: string, attributes: Record<string, string>) =>
  sqsService.setQueueAttributes(queueUrl, attributes)
export const changeMessageVisibility = (queueUrl: string, receiptHandle: string, visibilityTimeout: number) =>
  sqsService.changeMessageVisibility(queueUrl, receiptHandle, visibilityTimeout)

export default sqsService

File: pkg/proxy/internal/adapters/http/sqs.go
package httphandlers

import (
	"fmt"
	"net/http"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/sqs"
	"github.com/aws/aws-sdk-go-v2/service/sqs/types"
	"github.com/go-chi/chi/v5"
)

func (h *ProxyHandler) registerSQSRoutes(r chi.Router) {
	r.Route("/sqs", func(r chi.Router) {
		r.Get("/queues", h.listQueues)
		r.Post("/queues", h.createQueue)
		r.Get("/queues/{queueName}", h.getQueueUrl)
		r.Get("/queues/{queueName}/attributes", h.getQueueAttributes)
		r.Put("/queues/{queueName}/attributes", h.setQueueAttributes)
		r.Delete("/queues/{queueName}", h.deleteQueue)
		r.Post("/queues/{queueName}/messages", h.sendMessage)
		r.Get("/queues/{queueName}/messages", h.receiveMessage)
		r.Delete("/queues/{queueName}/messages/{receiptHandle}", h.deleteMessage)
		r.Post("/queues/{queueName}/purge", h.purgeQueue)
	})
}

func (h *ProxyHandler) listQueues(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &sqs.ListQueuesInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.SQS().ListQueues(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to list queues", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) createQueue(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &sqs.CreateQueueInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.SQS().CreateQueue(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to create queue", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) deleteQueue(w http.ResponseWriter, r *http.Request) {
	queueName := chi.URLParam(r, "queueName")

	// Look up the QueueUrl from the queue name
	urlResult, err := h.Svc.SQS().GetQueueUrl(h.ctx, &sqs.GetQueueUrlInput{
		QueueName: aws.String(queueName),
	})
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to get queue URL", err)
		return
	}

	input := &sqs.DeleteQueueInput{
		QueueUrl: urlResult.QueueUrl,
	}
	// Also try reading QueueUrl from body for backward compatibility
	if bodyBytes := readBody(r); len(bodyBytes) > 0 {
		var body struct {
			QueueUrl string `json:"QueueUrl"`
		}
		if err := parseBody(bodyBytes, &body); err == nil && body.QueueUrl != "" {
			input.QueueUrl = aws.String(body.QueueUrl)
		}
	}

	result, err := h.Svc.SQS().DeleteQueue(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to delete queue", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) getQueueUrl(w http.ResponseWriter, r *http.Request) {
	input := &sqs.GetQueueUrlInput{
		QueueName: aws.String(chi.URLParam(r, "queueName")),
	}
	result, err := h.Svc.SQS().GetQueueUrl(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to get queue URL", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) resolveQueueURL(w http.ResponseWriter, r *http.Request) *string {
	queueName := chi.URLParam(r, "queueName")
	urlResult, err := h.Svc.SQS().GetQueueUrl(h.ctx, &sqs.GetQueueUrlInput{
		QueueName: aws.String(queueName),
	})
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to get queue URL", err)
		return nil
	}
	return urlResult.QueueUrl
}

func (h *ProxyHandler) sendMessage(w http.ResponseWriter, r *http.Request) {
	queueURL := h.resolveQueueURL(w, r)
	if queueURL == nil {
		return
	}
	bodyBytes := readBody(r)
	input := &sqs.SendMessageInput{
		QueueUrl: queueURL,
	}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.SQS().SendMessage(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to send message", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) receiveMessage(w http.ResponseWriter, r *http.Request) {
	queueURL := h.resolveQueueURL(w, r)
	if queueURL == nil {
		return
	}
	input := &sqs.ReceiveMessageInput{
		QueueUrl: queueURL,
	}
	// Read optional params from query string
	if v := r.URL.Query().Get("MaxNumberOfMessages"); v != "" {
		var m int
		if _, err := fmt.Sscanf(v, "%d", &m); err == nil && m > 0 {
			input.MaxNumberOfMessages = int32(m)
		}
	}
	if v := r.URL.Query().Get("WaitTimeSeconds"); v != "" {
		var m int
		if _, err := fmt.Sscanf(v, "%d", &m); err == nil && m > 0 {
			input.WaitTimeSeconds = int32(m)
		}
	}
	if v := r.URL.Query().Get("VisibilityTimeout"); v != "" {
		var m int
		if _, err := fmt.Sscanf(v, "%d", &m); err == nil && m > 0 {
			input.VisibilityTimeout = int32(m)
		}
	}
	// Also try reading from body for backward compatibility
	if bodyBytes := readBody(r); len(bodyBytes) > 0 {
		var body struct {
			MaxNumberOfMessages int32 `json:"MaxNumberOfMessages"`
			WaitTimeSeconds     int32 `json:"WaitTimeSeconds"`
			VisibilityTimeout   int32 `json:"VisibilityTimeout"`
		}
		if err := parseBody(bodyBytes, &body); err == nil {
			if body.MaxNumberOfMessages > 0 {
				input.MaxNumberOfMessages = body.MaxNumberOfMessages
			}
			if body.WaitTimeSeconds > 0 {
				input.WaitTimeSeconds = body.WaitTimeSeconds
			}
			if body.VisibilityTimeout > 0 {
				input.VisibilityTimeout = body.VisibilityTimeout
			}
		}
	}
	result, err := h.Svc.SQS().ReceiveMessage(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to receive message", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) deleteMessage(w http.ResponseWriter, r *http.Request) {
	queueURL := h.resolveQueueURL(w, r)
	if queueURL == nil {
		return
	}
	input := &sqs.DeleteMessageInput{
		QueueUrl:      queueURL,
		ReceiptHandle: aws.String(urlParam(r, "receiptHandle")),
	}
	result, err := h.Svc.SQS().DeleteMessage(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to delete message", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) purgeQueue(w http.ResponseWriter, r *http.Request) {
	queueURL := h.resolveQueueURL(w, r)
	if queueURL == nil {
		return
	}
	input := &sqs.PurgeQueueInput{
		QueueUrl: queueURL,
	}
	result, err := h.Svc.SQS().PurgeQueue(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to purge queue", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) getQueueAttributes(w http.ResponseWriter, r *http.Request) {
	queueName := chi.URLParam(r, "queueName")

	// Look up the QueueUrl from the queue name
	urlResult, err := h.Svc.SQS().GetQueueUrl(h.ctx, &sqs.GetQueueUrlInput{
		QueueName: aws.String(queueName),
	})
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to get queue URL", err)
		return
	}

	input := &sqs.GetQueueAttributesInput{
		QueueUrl: urlResult.QueueUrl,
	}

	// Read optional AttributeNames from query params
	if attrNames := r.URL.Query()["AttributeName"]; len(attrNames) > 0 {
		for _, n := range attrNames {
			input.AttributeNames = append(input.AttributeNames, types.QueueAttributeName(n))
		}
	}

	result, err := h.Svc.SQS().GetQueueAttributes(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to get queue attributes", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) setQueueAttributes(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &sqs.SetQueueAttributesInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.SQS().SetQueueAttributes(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to set queue attributes", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}
